bundles/org.openhab.binding.visualcrossing/src/main/java/org/openhab/binding/visualcrossing/internal/api/rest/RetryHttpClient.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.visualcrossing.internal.api.rest;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.visualcrossing.internal.api.VisualCrossingApiException;
+import org.openhab.binding.visualcrossing.internal.api.VisualCrossingAuthException;
+import org.openhab.binding.visualcrossing.internal.api.VisualCrossingRateException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Martin Grześlowski - Initial contribution
+ */
+@NonNullByDefault
+public class RetryHttpClient implements RestClient {
```

Comment:
Did you consider to integrate `RetryHttpClient` and `HttpClient` ? As the only real difference is setting `maxRetries` to 0 or 3.

Reply:
No, you can have HttpClient without retries

Reply:
> No, you can have HttpClient without retries

Wouldn't that in essence be `RetryHttpClient` with 0 retries ? It would de-duplicate code and make it (a little bit) simpler.

Reply:
Check how `execute` method would look like if it had both retry and http query:

```java
    @SuppressWarnings("ConstantValue")
    private @Nullable String execute(Request request, @Nullable Header[] headers, String url)
            throws VisualCrossingApiException, VisualCrossingAuthException, VisualCrossingRateException {
        for (int i = 0; i < 10; i++) {
            try {
                try {
                    if (headers != null) {
                        for (var header : headers) {
                            if (header == null) {
                                continue;
                            }
                            for (var value : header.values()) {
                                request.header(header.name(), value);
                            }
                        }
                    }
                    request.timeout(TIMEOUT, SECONDS);
                    request.idleTimeout(IDLE_TIMEOUT, SECONDS);
                    var response = request.send();
                    var status = response.getStatus();
                    if (status < 200 || status >= 399) {
                        throw new HttpVisualCrossingApiException(status, response.getReason());
                    }
                    return response.getContentAsString();
                } catch (RuntimeException | TimeoutException | ExecutionException | InterruptedException ex) {
                    Throwable cause = ex;
                    while (cause != null) {
                        if (cause instanceof HttpResponseException hte) {
                            var response = hte.getResponse();
                            int status = response.getStatus();
                            if (status == 401) {
                                throw new VisualCrossingAuthException();
                            }
                            if (status == 429) {
                                throw new VisualCrossingRateException();
                            }
                            throw new HttpVisualCrossingApiException(response.getStatus(), response.getReason(), hte);
                        }
                        cause = cause.getCause();
                    }
                    throw new VisualCrossingApiException("Error while executing request to " + url, ex);
                }
            } catch (VisualCrossingApiException ex) {
                if (i < 10 - 1) {
                    logger.debug("Error while calling GET {}. Retrying {}/{}...", url, i + 1, 10, ex);
                } else {
                    throw ex;
                }
            }
        }
        throw new IllegalStateException();
    }
```

Reply:
There are multiple implementations possible, but as the current one is not _wrong_  i'll set this to resolved.